### PR TITLE
Re-based fix-837 branch to master with fixed-changelog.

### LINKED ISSUE #837: CancelScope shielding fails to protect against cancellation until task_status.started is called on asyncio
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

4.7.0

### Python version

3.10

### What happened?

Running some quite involved code that needs to do shielded cleanup if it fails sometime exits a shielded scope due to external cancellation if it was cancelled before it had called `task_status.started()` 

I know these are edge cases and may also be really hard to solve so feel free to resolve as such.

### How can we reproduce the bug?

I've managed to reduce it to the following test-case.

```python

async def test_anyio_cancel_shielding_start():
    entered_inner_scope = anyio.Event()
    async with (
        anyio.create_task_group() as tg,
        anyio.create_task_group() as inner_tg,
    ):
        async def inner_task(*, task_status: anyio.abc.TaskStatus = anyio.TASK_STATUS_IGNORED):
            # task_status.started() # Uncommenting this lets the task hang forever instead

            with anyio.CancelScope(shield=True):
                entered_inner_scope.set()
                try:
                    await anyio.sleep_forever()
                except anyio.get_cancelled_exc_class():
                    assert False, "Shouldn't be cancelled in a shielded scope"

        async def start_inner_task():
            await inner_tg.start(inner_task)

        tg.start_soon(start_inner_task)
        await entered_inner_scope.wait()
        tg.cancel_scope.cancel()



if __name__ == "__main__":
    anyio.run(test_anyio_cancel_shielding_start)
```

### PR DESCRIPTION
**Pull request by tapetersen**
**NOTE** Erasing or replacing the contents of this template will result in your pull
request being summarily closed without consideration!

## Changes

Fixes #837 (or rather as noted just adds a regression test and note in changelog for it as it was already solved>
This otherwise just a rebase of #839

### CODE REVIEW

**Comment by tapetersen** [comment]
@agronholm I figured the changelog entry that it's fixed is nice to have for confirming version and the test doesn't hurt.
Feel free to close otherwise.

File: docs/versionhistory.rst [hunk 1]
```diff
@@ -3,6 +3,16 @@ Version history
...
+**UNRELEASED**
+
+- Fixed cancellation edge case on asyncio where a task spawning another with
+  ``TaskGroup.start()`` is not protected from external cancellation even when
+  the subtask has not yet called ``task_status.started()`` and is in a shielded
+  cancel scope.
+  This was actually solved already in 4.14.0 by ``TaskHandle`` refactoring.
+  This patch just adds a regression test.
+  (`#837 <https://github.com/agronholm/anyio/issues/837>`_)
+
```
Comment (agronholm) [inline docs/versionhistory.rst]: An already fixed problem nor an added test warrant a changelog entry.

Comment (tapetersen) [reply docs/versionhistory.rst]: Corrected

File: tests/test_taskgroups.py [hunk 2]
```diff
@@ -2008,6 +2008,34 @@ async def test_cancel_reason() -> None:
...
+
+        # The cancellation should be triggered here, and not any earlier
+        await checkpoint()
+
+    entered_inner_scope = Event()
+    with CancelScope() as outer_scope:
+        async with create_task_group() as tg:
+            tg.start_soon(tg.start, taskfunc)
+            await entered_inner_scope.wait()
+            outer_scope.cancel()
```
Comment (agronholm) [inline tests/test_taskgroups.py]: I feel these changes would make the test less reliant on the scheduling order. What do you think?
```suggestion
                await resume_event.wait()
            except get_cancelled_exc_class():
                pytest.fail("Shouldn't be cancelled in a shielded scope")

        # The cancellation should be triggered here, and not any earlier
        await checkpoint()

    resume_event = Event()
    with CancelScope() as outer_scope:
        async with create_task_group() as tg:
            tg.start_soon(tg.start, taskfunc)
            await wait_all_tasks_blocked()
            outer_scope.cancel()
            resume_event.set()
```

Comment (tapetersen) [reply tests/test_taskgroups.py]: Absolutely! I'm all for explicit ordering for tests.

(I think it was you that added checkpoints  when transferring my initial testcase and i thought about it but let it be as a simple matter of taste between consiceness of test-case (I very rarely used them at all at that point and like I said I tend to prefer explicit Events whenever I write tests myself)).

**Comment by tapetersen** [comment]
@agronholm Your refactoring fixed this case but there are still some less serious differences from Trio. I was unsure if you though they were deemed serious enough to file Issues and potential fixes for. If they are I'll do with proper demonstrators.

Since we now start the `Task` directly in the `TaskGroup` rather than move it in when `started()` is called as Trio there are at least 3 visible side-effects if the `TaskGroup` is cancelled in that timeframe (and we assume `TaskGroup.start` is called by a not descending itself from that group/scope):

1. On Trio the started task doesn't get a `CancelledError` until the first checkpoint *after* the `started()` call.
2. Related to the first on Trio that means any value passed to `started` will be available and correctly passed back to the caller of `start`
3. On the asyncio backend the `CancelledError` leaks out of the `start()` call to the calling task (the most serious).

Our work is potentially affected by the last (3) as in it will potentially lead to a more confusing debug experience and possibly has already, still not critical though. It's also the one that could probably be relatively simply fixed with a workaround while 1 and 2 would require some more refactoring from what I can see.

**Comment by agronholm** [comment]
If a piece of code works differently on two different back-ends, then it's almost certainly a bug (except for task scheduling differences, cancellation exception differences and stuff like that of course).

**Review by agronholm — APPROVED** [review]
